### PR TITLE
Issue #2127: Allow user override default SASL service name bookkeeper

### PR DESCRIPTION
**Pull request by 29x10**
Descriptions of the changes in this PR:
default SASL service name "bookkeeper" can be override by JVM property "bookkeeper.sasl.servicename"

### Changes

Instead of use a constant value, it would read from JVM property first, if it doesn't exists, then use default value from constant variable SaslConstants.SASL_BOOKKEEPER_PROTOCOL

Master Issue: #2127

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java11]: skip build on java11. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslClientState.java [hunk 1]
```diff
@@ -50,7 +50,9 @@
     private String password;
 
     public SaslClientState(String serverHostname, Subject subject) throws SaslException {
-        String serverPrincipal = SaslConstants.SASL_BOOKKEEPER_PROTOCOL + "/" + serverHostname;
+        String saslServiceName = System.getProperty("bookkeeper.sasl.servicename",
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslClientState.java]: This should be configurable in Client Configuration and in ServerConfiguration.

We should add tests as well, you can check existing tests and start from them

Comment (29x10) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslClientState.java]: Sorry I couldn't find any existing tests that use SaslClientState.

And this only affect client side, when client authenticate with server, you need to provide server's service name.

Comment (jiazhai) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslClientState.java]: @eolivelli, We will also add some other JVM properties for sasl, using system property is fine for me.  Comparing with add in `Client Configuration`, this change made less changes. 
@29x10 FYI. `GSSAPIBookKeeperTest` is the test case.

Comment (29x10) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/sasl/SaslClientState.java]: @jiazhai I just add the test case, please help to review

**Comment by jiazhai** [comment]
rebuild java11
rebuild java8
run integration tests

**Comment by jiazhai** [comment]
rebuild java11
rebuild java8
run integration tests

**Comment by sijie** [comment]
@29x10 can you rebase the pull request to merge the latest master? the CI was failing due to #2125. #2125 has been merged to latest master.

**Comment by 29x10** [comment]
run bookkeeper-server replication tests

**Comment by 29x10** [comment]
@eolivelli Could you help to take a look at the change with tests?

**Comment by eolivelli** [comment]
I will take a new look tomorrow. I can't remember all the configurable parameters as system props now, I will check the code as soon as possible.
Sorry for the delay

**Comment by 29x10** [comment]
@eolivelli There is no rush, please take your time

**Comment by 29x10** [comment]
@eolivelli 

1. I agree with you to define it in SaslConstants.
2. Why I use system properties is because zookeeper use it too, "zookeeper.sasl.client.username".
3. Yes, it disallows you connect to different BK clusters if they use different service name in the name process.
4. Definitely I would update documentation page as well if you are fine with my changes.

**Comment by eolivelli** [comment]
@29x10 please update the doc and I will merge this change, as @sijie already approved it
we are going to cut a release soon

**Comment by 29x10** [comment]
@eolivelli Please review the changes, I update the docs to `site/docs/4.9.2/security/sasl.md` and add the property name to SaslConstants.

**Comment by sijie** [comment]
run bookkeeper-server replication tests

**Comment by 29x10** [comment]
@eolivelli I rebased with master branch, due to issue #2133 , it fails.

**Comment by eolivelli** [comment]
let's wait for travis, then I will merge this patch

thank you

**Pull request merged**